lib/puppet/type/nssdb_certificate.rb

```diff
@@ -0,0 +1,52 @@
+Puppet::Type.newtype(:nssdb_certificate) do
+  desc 'adds a certificate to an nssdb'
+
+  ensurable
+
+  def self.title_patterns
+    [ [ /(.+):(.+)/m, [ [:nssdb], [:cert_name] ] ] ]
+  end
+
+  newparam(:cert_name, :namevar => true) do
+    desc "The certificate name used to store inside the nssdb"
+  end
+
+  newparam(:nssdb, :namevar => true) do
+    desc "Path to the nssdb to use or create when importing the certificate"
+    isrequired
+  end
+
+  newproperty(:certificate) do
+    desc "Path to the certificate to add to the nssdb"
+
+    def fingerprint(file)
+      provider.fingerprint(file)
+    end
+
+    def should_to_s(newvalue)
+      self.class.format_value_for_display(fingerprint(newvalue))
+    end
+
+    def insync?(is)
+      is == fingerprint(should)
+    end
+  end
+
+  newparam(:private_key) do
+    desc "Path to the private key to add to the nssdb"
+  end
+
+  newparam(:trustargs) do
+    desc "Certificate trust flags for certificate inside the nssdb"
+    isrequired
+  end
+
+  newparam(:password_file) do
+    desc "Path to file containing the nssdb password"
+    isrequired
+  end
+
+  autorequire(:file) do
+    [self[:password_file], File.dirname(self[:nssdb]), self[:certificate], self[:private_key]]
```

Comment:
Should this require the nssdb directory itself? That can ensure the directory is never world readable: first ensure the directory has correct permissions, then run nssdb commands.

Reply:
Sorry, I do not follow the nuance of the question. `File.dirname(self[:nssdb])` causes a require on the directory.

Reply:
The parent directory, but not the directory itself. That can mean you end up running
* `File.dirname(self[:nssdb])`
* nssdb_certificate itself
* `self[:nssdb]`

What I'm asking is: should we always ensure:
* `self[:nssdb]`
* nssdb_certificate itself

And because the file type also requires parents, you get `File.dirname()` for free.

Reply:
Ahh yes, will make that change. A bit of a copy paste fail on my part adapting the truststore_certificate work.